### DISCUSSION TITLE
Two invalidate make `data: null`, is this a bug?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#2417] [Q&A] [answered]
I have this updater:

```ts
const updates: GraphCacheUpdaters = {
  Subscription: {
    updatedOrDeleted(parent, _, cache) {
      const { resourceType, resourceID } = parent.updatedOrDeleted;
      cache.invalidate({ __typename: resourceType, id: resourceID });
    },
  },
};
```

It worked amazingly, until today.

In a schema like this:

```graphql
type User {
  id: ID!
  name: String
  ProfileID: ID!
  Profile: Profile
}
```

if the backend sends two actions immediately one after the other:

```
updatedOrDeleted: {action: "UPDATED", resourceType: "User", resourceID: "1"}
updatedOrDeleted: {action: "DELETED", resourceType: "Profile", resourceID: "1"}
```

urql make my users list `data: null`.

Why?

There are problems with two invalidate one after another?

```
"@urql/core": "2.4.4",
"@urql/devtools": "2.0.3",
"@urql/exchange-graphcache": "4.3.6",
"@urql/exchange-persisted-fetch": "1.3.4",
"@urql/exchange-retry": "0.3.3",
"@urql/svelte": "1.3.3",
```

**Comment by JoviDeCroock** [comment, marked as answer]
Because your list has [User!] probably, unlink might be better for non-nullable lists

**Reply by frederikhors** [reply]
@JoviDeCroock thanks a ton for you answer.

With `unlink` you mean https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#updating-many-unknown-links?

**Reply by frederikhors** [reply]
Yes, my list comes from this:

```graphql
type UserListConnection {
  nodes: [User!]!
  info: ConnectionInfo!
}
```

**Reply by frederikhors** [reply]
I opened https://github.com/FormidableLabs/urql/issues/2418 with a reproduction.

**Reply by JoviDeCroock** [reply]
I mean, the explanation is there, you invalidate a non-nullable entity so we bubble null up

**Reply by frederikhors** [reply]
I do not understand. What is the non-nullable entity?

**Reply by JoviDeCroock** [reply]
User because of the ! So invalidate on the nullable profile --> partial data else we refetch the users as we see a null in your array which is not allowed

**Reply by frederikhors** [reply]
Can you please talk about the schema in this repro: https://stackblitz.com/edit/sveltejs-kit-template-default-9dvq26?file=server%2Fschema.js

I'm still not understanding the cause...

**Reply by frederikhors** [reply]
And if there is a way to fix... I'm desperate.

**Reply by frederikhors** [reply]
I mean the `invalidate` as a way to re-fetch the object. Am I wrong?

**Reply by frederikhors** [reply]
@JoviDeCroock I updated the repro in https://github.com/FormidableLabs/urql/issues/2418 adding nullable array:


```graphql
type Query {
  games: [Game!]!
  players: [Player]
}
```

but the issue is still there.

**I'm desperate...** 😢

**Reply by kitten** [reply]
I'm not sure what you're expecting there since I don't know what this app does and don't have time to really read all of it.
But if I'm not misunderstanding it, your updater is called on a subscription and then does a `cache.invalidate` call, correct? How would graphcache recover from that given that there's no schema input?

If you're deleting data from the cache then a new response will be generated. However, when you're deleting data then that means that the updater will simply delete the data and trigger a new background fetch call. Is there something unclear about that? or, what specifically are you expecting?

Also, I'd love if you could keep this to the dicussion at #2417 please. Duplicates are simply not appropriate when you're asking a question about any of the libraries' behaviour. I'll leave it up to you which of the two you'd like to use for the question itself, but one will have to be closed

**Reply by frederikhors** [reply]
Thank you for your answer @kitten.

I'll close the issue if we can talk here, no problem. But please do not abandon me...

> I'm not sure what you're expecting

I need to update those data in real-time with subscriptions.

I expect `invalidate` to re-fetch those queries. Stop. No more.

And it does. On one browser tab it works almost every time. Suddendly also on one tab (with the same operations add/remove) `data` becomes `null`.

> How would graphcache recover from that given that there's no schema input?

On my backend I have all setup good: schema (for awareness), local resolvers for list > detail, everything and the problem is the same.

**I simply can't understand** why in the other tab (or suddendly in the same one) players array or games array become `null` INSTEAD of a simple refetch.

### One tab only problem (it worked many times before this one delete click):

![issDataNull](https://user-images.githubusercontent.com/41120635/164912481-35f2225f-535f-41fd-be36-a039bd71b4e9.gif)

**Reply by kitten** [reply]
I can't really get the sandbox to run, since it just leads to CORS issues and other problems.
However, what I'd need to see is `debugExchange` output, since the above behaviour doesn't quite seem to plausibly show us the full picture since `stale: false` is present.

**Reply by frederikhors** [reply]
Can I help with the cors issue? There are no cors problem since it's all on stackblitz, server & client.

**Reply by frederikhors** [reply]
How can I help you debugging?

**Reply by frederikhors** [reply]
Can you download the code using the download button?

![image](https://user-images.githubusercontent.com/41120635/164934012-b9afacb7-239b-4a82-9a89-4fd592ca853f.png)

**Reply by frederikhors** [reply]
I added `debugExchange`, this happens on that delete click:

![image](https://user-images.githubusercontent.com/41120635/164934189-76c4256e-bdd8-4853-a21c-d18c9721e46a.png)

**Reply by frederikhors** [reply]
I changed the reproduction: you do not have to do anything now: just open the browser at: https://stackblitz.com/edit/sveltejs-kit-template-default-9dvq26.

**Reply by frederikhors** [reply]
Are you still with me, @kitten? 😢

**Reply by JoviDeCroock** [reply]
Frederik, sorry that you are encountering an issue but this repository is not our job and it's weekend for us as well. If you want something urgently fixed you can always fork and fix it yourself

**Reply by frederikhors** [reply]
I know. I have always been respectful and have always tried to help as much as I could.

It is that this problem is urgent because it was discovered in production.

Excuse me. I'll await calmly. Sorry.
